plugin/src/frontend.ml

```diff
@@ -34,27 +34,43 @@ let maybe_prove_coherence n inv_n idx_n : unit =
     Printf.printf "Defined coherence proof %s\n\n" (Id.to_string coh_n)
   else
     ()
-      
+
 (*
- * If the option is enabled, then prove section and retraction after find_ornament is called.
- * Otherwise, do nothing.
+ * If the option is enabled, then prove section, retraction, and adjunction after
+ * find_ornament is called. Otherwise, do nothing.
  *)
 let maybe_prove_equivalence n inv_n : unit =
+  let define_proof suffix ?(adjective=suffix) evd term =
+    let ident = with_suffix n suffix in
+    let const = define_term ident evd term true |> destConstRef in
+    Printf.printf "Defined %s proof %s\n\n" adjective (Id.to_string ident);
+    const
+  in
   if is_search_equiv () then
     let env = Global.env () in
     let evd = Evd.from_env env in
     let (promote, forget) = map_tuple make_constant (n, inv_n) in
     let l = initialize_lifting env evd promote forget in
     let (section, retraction) = prove_equivalence env evd l in
-    let sec_n = with_suffix n "section" in
-    let _ = define_term sec_n evd section true in
-    Printf.printf "Defined section proof %s\n\n" (Id.to_string sec_n);
-    let rec_n = with_suffix n "retraction" in
-    let _ = define_term rec_n evd retraction true in
-    Printf.printf "Defined retraction proof %s\n\n" (Id.to_string rec_n)
+    let sect = define_proof "section" evd section in
+    let retr0 = define_proof "retraction0" evd retraction ~adjective:"pre-retraction" in
+    let pre_adjoint = { orn = l; sect; retr0 } in
+    let _ =
+      let env = Global.env () in
```

Comment:
Clarify explicitly in the code why we need a new environment and evar_map: We have just defined some terms, so we need the updated global environment that has those terms, and the corresponding evar_map with updated constraints. Can just be a comment if you'd like, or you can have a "refresh_environment" function that defines a new global env and gets a new evar_map for that env, with its name serving as documentation.

Reply:
Done. I just inserted comments.